### PR TITLE
[v1.1]Introduce Wasp Agent Eviction Logic and Initial Functional Test

### PR DESCRIPTION
**Pull request by Barakmor1**
This pull request introduces significant enhancements to the Wasp agent, focusing on the eviction logic and adding the initial functional test. The changes are inspired by best practices and structures from the kubevirt/kubevirt repository.

Key additions:
1. Infrastructure Setup for Golang Rewrite:

- Makefile:

    - `make manifests` rule to generate Wasp agent manifests in the _out directory.
    - `make cluster-up` to provision a Kubernetes cluster with swap enabled using kubevirtci. Set `export KUBEVIRT_PROVIDER=k8s-1.30` before deploying the cluster.
    - `make cluster-sync` to build and deploy Wasp agent images locally onto the cluster with code changes. Set `export KUBEVIRT_PROVIDER=external` for deployment on an external cluster with a configured KUBECONFIG (in this case, `make cluster-up` is not needed).
    - `make test` for running unit tests.
    -  `make functest` for executing functional tests.

2. Eviction Logic Components:
- Application Setup:
    - `pkg/application.go`: Application setup logic.
    - `pkg/application_test.go`: Examples for unit tests.
- Informers Setup:
    - `pkg/informers`: Shared informers setup.
- Wasp Client:
    - `pkg/client`: Wasp client implementation.
- Logging:
    - `pkg/log`: Package for logging.
- Resources:
    - `pkg/resources`: Manifests for autogeneration with make manifests.

3. Eviction Logic Specifics:

- Pod Evictor:
    - `pkg/wasp/pod-evictor`: Logic for forced pod deletion.
- Pod Ranker:
    - `pkg/wasp/pod-ranke`r: Logic for selecting pods to evict (currently WIP, using a dummy ranker to choose pods with names containing the word 'memory').
- Pod Filter:
    - `pkg/wasp/pod-filter`: Logic for filtering pods that cannot be evicted, such as static pods and those in namespaces with 'openshift' or 'kube-system' prefixes, or the Wasp installation namespace.
- Stats Collector:
    - `pkg/wasp/stats-collector`: Logic for gathering node stats (swap in/out and available memory) every second to share with the shortage detector.
- Shortage Detector:
    - `pkg/wasp/shortage-detector`: Logic for monitoring node stats and triggering eviction during shortages.
- Eviction Controller:
    - `pkg/wasp/eviction-controller`: Logic invoked every 5 seconds to check for shortages and perform evictions using the above components.
    
4. Initial Functional Test:

- Test Scenario:
    - Two pods are added to a schedulable node: one 'innocent' pod with minimal memory usage, and another pod that stresses the node by adding 300 MiB every 10 seconds.
A logger thread monitors swap in/out and available memory of the selected node every 3 seconds, ensuring the eviction logic functions as expected under various conditions.

5. Tunning
- It is possible to tune parameters before deploying the Wasp agent. Here are the tuning environment variables that can be set before running make `cluster-sync` / `make manifests` and affect the eviction:
    - `MAX_AVERAGE_SWAPIN_PAGES_PER_SECOND` (default: 1000)
    - `MAX_AVERAGE_SWAPOUT_PAGES_PER_SECOND` (default: 1000)
    - `AVERAGE_WINDOW_SIZE_SECONDS` (default: 30)
    - `MEMORY_OVER_COMMITMENT_THRESHOLD` (default: 500Mi)

### CODE REVIEW

File: Containerfile.wasp [hunk 1]
```diff
@@ -0,0 +1,34 @@
+# Builder stage
+FROM golang:1.20-alpine as builder
```
Comment (fabiand) [inline Containerfile.wasp:2]: Please use fedora as a base.

File: README.md [hunk 2]
```diff
@@ -1,41 +1 @@
...
-
-    $ oc apply -f manifests/static.yaml
-    $ oc apply -f manifests/stress.yaml
-    # Now increase the stress deployment replicacount in order to push out
-    # memory of static pods
-
-# Destroy
-
-    $ bash to.sh destroy
+# wasp
```
Comment (fabiand) [inline README.md:1]: Please add some meaningful content here.

File: cluster-sync/clean.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
+#!/bin/bash -e
```
Comment (fabiand) [inline cluster-sync/clean.sh:1]: Could cluster-sync be a submodule?

File: cluster-up/check.sh [hunk 4]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bash
```
Comment (fabiand) [inline cluster-up/check.sh:1]: could cluster-up be a git submodule?

Comment (enp0s3) [reply cluster-up/check.sh:1]: @fabiand I had bad times working with sub-modules. I would suggest to keep it as is for now, and later when we will have CI, we can always bump it the same way as we do in KV.

File: Containerfile.wasp (hunk 1 above)
Comment (Barakmor1) [reply Containerfile.wasp:2]: Can you please elaborate why is it better?

File: README.md (hunk 2 above)
Comment (Barakmor1) [reply README.md]: Done

File: Containerfile.wasp (hunk 1 above)
Comment (fabiand) [reply Containerfile.wasp:2]: Because we are building for openshift, and openshift is based on rhel, and rhel on fedora.

**Review by enp0s3 — DISMISSED** [review]

**Review by enp0s3 — DISMISSED** [review]

**Review by enp0s3 — APPROVED** [review]

File: Containerfile.wasp (hunk 1 above)
Comment (fabiand) [reply Containerfile.wasp:2]: Has this been addressed?

**Review by fabiand — CHANGES_REQUESTED** [review]
This PR should be against `main`. Once it landed there, then we can branch off v1.1.

**Review by fabiand — CHANGES_REQUESTED** [review]
#40 has to be merged first.

And this PR has to be aligned to #40 once merged.

**Comment by stu-gott** [comment]
https://github.com/openshift-virtualization/wasp-agent/pull/40 has merged

**Comment by enp0s3** [comment]
@stu-gott @fabiand Can we close this PR and do the cut for `v1.1` close to the FF date?
we want to continue working against the main branch until then.

**Comment by fabiand** [comment]
@enp0s3 I'm absolutely in favor of that.

I'd even remove the v1.1 branch. and just cut it once feature freeze comes up.

**Pull request closed without merging**